<h2><font color="#004D7F" size=6> Procesamiento de textos  Tarea4</font></h2>



<br><br>
<div style="text-align: right">
<font color="#004D7F" size=3>Danyela Luengas</font><br>
<font color="#004D7F" size=3>Procesamiento de lenguaje natural o minería de textos</font><br>

Para el preprocesamiento de los textos puede:

1. Estandarizar el texto a minúsculas
2. Eliminar las menciones a usuarios (@user)
3. Eliminar las url’s
4. Eliminar los emojis
5. Las abreviaturas, contracciones y slangs sustituirlas por el texto equivalente
6. Eliminar palabras funcionales
7. Verificar si existen cifras numéricas, las cuales pueden ser reemplazadas por algún término o eliminarlas
8. Tratamiento con los hashtags
9. Eliminar caracteres raros y especiales
10. Eliminar signos de puntuación
11. Estandarizar las secuencias de varios espacios en blanco, tabuladores y saltos de línea
Entre otras…
Posibles características para tenerse en cuenta:

1. N-gramas de caracteres
2. N-gramas de palabras
3. N-gramas de etiquetas POS
4. N-gramas de saltos de palabras (skip-gram)
5. N-gramas de palabras funcionales
6. N-gramas de símbolos de puntuación
7. Entre otras…

In [2]:
import re
import pandas as pd
import numpy as np
import nltk
from nltk import *

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics import accuracy_score, f1_score

In [9]:
tr=pd.read_csv('train_es2.tsv',delimiter='\t', encoding='utf8')

des=pd.read_csv('dev_es2.tsv',delimiter='\t', encoding='utf8')


In [15]:
pattern_URL="(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9]\.[^\s]{2,})"

def procesar(file, namefile):    
    file[file.columns[1]] = [clean_text(i) for i in file[file.columns[1]]]    
    file.to_csv(namefile, sep='\t', encoding='utf-8', index=False)
    return file
    
def clean_text(text):
    text = text.lower()   #  convertir en minuscula
    #text=re.sub("@([A-Za-z0-9_]{1,15})", "@USUARIO", text)
    text=re.sub("@([A-Za-z0-9_]{1,15})", " ", text)
    text=re.sub(pattern_URL, " ", text)
    
    text= remove_emoji(text)
    text= remove_stopwords(text)
    text=re.sub("\d+", "0", text)
    # text=re.sub("\d+", " ", text)
    
    #elimnar SIGNOS DE PUNTUACIÓN
    text=re.sub(r" +", " ", re.sub(r"\t", " ", re.sub(r"\n+", "\n", re.sub('(?:[.,\/!$%?¿?!¡\^&\*;:{}=><\-_`~()”“"\'\|])', " ",text))))
    text = text.strip()
    return text
#
def remove_stopwords(text):    
    
    stopwords=set(nltk.corpus.stopwords.words("spanish"))
    for i in stopwords:
        text = re.sub(r"\b%s\b" % i, " ", text)
    return text

def remove_emoji(text):
    emoji_pattern = re.compile("["
                               "\U0001F600-\U0001F64F"  # emoticons
                               "\U0001F300-\U0001F5FF"  # symbols & pictographs                               
                               "\U0001F680-\U0001F6FF"  # transport & map symbols
                               "\U0001F1E0-\U0001F1FF"  # flags (iOS)
                               "\U00002702-\U000027B0"
                               "\U000024C2-\U0001F251"
                               "\U0001f926-\U0001f937"
                               "\u200d"
                               "\u2640-\u2642"
                               "\U0001F1F2-\U0001F1F4"  # Macau flag
                               "\U0001F1E6-\U0001F1FF"  # flags
                               "\U0001F600-\U0001F64F"
                               "\U0001F1F2"
                               "\U0001F1F4"
                               "\U0001F620"
                               "]+", flags=re.UNICODE)   
    text = emoji_pattern.sub(r'', text) # no emoji
    return text

['id', 'text', 'hs', 'tr', 'ag', 'easyjet', 'quiere', 'duplicar', 'el', 'número', 'de', 'mujeres', 'piloto', 'verás', 'tú', 'para', 'aparcar', 'el', 'avión', 'http', 'el', 'gobierno', 'debe', 'crear', 'un', 'control', 'estricto', 'de', 'inmigración', 'en', 'las', 'zonas', 'fronterizas', 'con', 'colombia', 'por', 'q', 'después', 'del', 'venir', 'en', 'masa', 'yo', 'veo', 'mujeres', 'destruidas', 'por', 'acoso', 'laboral', 'callejero', 'otras', 'con', 'depresión', 'debido', 'violación', 'sexual', 'maltrato', 'físico', 'conocí', 'varias', 'que', 'se', 'suicidaron', 'por', 'este', 'tipo', 'de', 'comportamientos', 'machistas', 'vas', 'seguir', 'con', 'tu', 'show', 'pobre', 'https', 'yo', 'soy', 'respetuoso', 'con', 'los', 'demás', 'sólamente', 'les', 'recuerdo', 'lo', 'que', 'son', 'si', 'son', 'escoria', 'es', 'culpa', 'mía', 'que', 'lo', 'sé', 'quién', 'tomas', 'antonio']


In [16]:
Train=procesar(des, 'train_es_clean2.tsv')
dev=procesar(tr, 'dev_es_clean2.tsv')
Train

id                                               text  HS  TR  AG
0    20005                    comiendo picada árabe rica vida   0   0   0
1    20006                            callate zorra mama duro   1   1   1
2    20011  acabo escuchar casado diciendo decenas millone...   0   0   0
3    20019  nadie preguntado va costar hombre guantazo sit...   1   0   1
4    20033                           callate come sobra zorra   1   1   1
..     ...                                                ...  ..  ..  ..
495  24930  habe pasame video adan diga puta mujeres sevil...   1   0   1
496  24953                            callate hijo perra amor   0   0   0
497  24973  auténtico imbécil potencial machista ideología...   0   0   0
498  24983  callate q paso mili m paso ami faltas vos gord...   1   1   1
499  24986  traducción escucha puta sueca árabes aquí apod...   1   1   1

[500 rows x 5 columns]

In [47]:
train = Train[Train.columns[0]]
X_train_textA = Train[Train.columns[1]].fillna(' ')
y_train_hsA = Train[Train.columns[2]]

test = dev[dev.columns[0]]
X_test_textA = dev[dev.columns[1]].fillna(' ')
y_test_hsA = dev[dev.columns[2]]
train

0      20005
1      20006
2      20011
3      20019
4      20033
       ...  
495    24930
496    24953
497    24973
498    24983
499    24986
Name: id, Length: 500, dtype: int64

In [17]:
txt = "Company10"

x = txt.isalpha()

print(x)


False


In [21]:
print( len(X_train_textA), len(y_train_hsA) )

print( len(X_test_textA), len(y_test_hsA) )


500 500
4469 4469


In [32]:
cvectorizer = CountVectorizer(
    # lowercase=True,
    #stop_words=[word.decode('utf-8') for word in nltk.corpus.stopwords.words('spanish')],
    #token_pattern=r'\b\w+\b', #selects tokens of 2 or more alphanumeric characters 
    ngram_range=(1,3),#n-grams de palabras n = 1 a n = 3 (unigramas, bigramas y trigramas)
    min_df=5,#ignorando los términos que tienen una frecuencia de documento estrictamente inferior a 5
).fit(X_train_textA)

X_train_cvectorized = cvectorizer.transform(X_train_textA).toarray()
print(X_train_cvectorized.shape)

X_test_cvectorized = cvectorizer.transform(X_test_textA).toarray()
print(X_test_cvectorized.shape)
y_train_hsA

(500, 152)
(4469, 152)


0      0
1      1
2      0
3      1
4      1
      ..
495    1
496    0
497    0
498    1
499    1
Name: HS, Length: 500, dtype: int64

In [30]:
tvectorizer = TfidfVectorizer(
    # lowercase=True,
    #stop_words=[word.decode('utf-8') for word in nltk.corpus.stopwords.words('spanish')],
    #token_pattern=r'\b\w+\b', #selects tokens of 2 or more alphanumeric characters 
    ngram_range=(1,3),#n-grams de palabras n = 1 a n = 3 (unigramas, bigramas y trigramas)
    min_df=5,#ignorando los términos que tienen una frecuencia de documento estrictamente inferior a 5
).fit(X_train_textA)

X_train_tvectorized = tvectorizer.transform(X_train_textA).toarray()
print(X_train_tvectorized.shape)

X_test_tvectorized = tvectorizer.transform(X_test_textA).toarray()
print(X_test_tvectorized.shape)


y_test_hsA

(500, 152)
(4469, 152)


In [31]:
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler(feature_range=(0, 1))
X_train = scaler.fit_transform(X_train_cvectorized)
X_test = scaler.fit_transform(X_test_cvectorized)


In [46]:
from sklearn.metrics import confusion_matrix,accuracy_score, classification_report, roc_curve

from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import mean_squared_error, r2_score

# Compare Algorithms
# prepare models
models=[]
RMSE = []
R_sq = []
models.append(('LOR', LogisticRegression(solver='lbfgs', max_iter=1000)))
models.append(('LDA', LinearDiscriminantAnalysis()))
models.append(('k-NN', KNeighborsClassifier(n_neighbors=3)))
models.append(('CART', DecisionTreeClassifier()))
models.append(('NB', GaussianNB()))
#models.append(('SVM', SVC(gamma='auto')))
models.append(('GBC',  GradientBoostingClassifier(n_estimators=200, learning_rate=0.7,
                                 max_depth=1)))


# evaluate each model in turn
results = []
names = []
scoring= 'accuracy'
for name, model in models:
    kfold= KFold(n_splits=10)
    cv_results= cross_val_score(model,X_train, y_train_hsA, cv=kfold, scoring=scoring)
    names.append(name)
    print(f'Accuracy de {name}: {cv_results.mean()*100.0:,.2f}% (std:{cv_results.std()*100.0:,.2f}%)')

Accuracy de LOR: 71.80% (std:6.60%)
Accuracy de LDA: 64.80% (std:6.34%)
Accuracy de k-NN: 69.60% (std:5.78%)
Accuracy de CART: 64.60% (std:4.20%)
Accuracy de NB: 67.00% (std:7.96%)
Accuracy de GBC: 71.00% (std:6.08%)


In [40]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import mean_squared_error, r2_score


GBC = GradientBoostingClassifier(n_estimators=200, learning_rate=1.0,
                                 max_depth=1)
GBC.fit(X_train_tvectorized, y_train_hsA)
predicGBC = GBC.predict(X_test)
print('La probabilidad de acierto  para GradientBoostingClassifier \n')
print((np.sum(predicGBC == y_test_hsA)/len(predicGBC))*100)
print(predicGBC.sum())
print('Matrix de Confusion \n')
print(pd.crosstab(y_test_hsA, predicGBC, rownames=['True'], colnames=['Predicted'], margins=True))
print(classification_report(y_test_hsA, predicGBC))

print("Accuracy con Datos test {:.2f}".format(accuracy_score(y_test_hsA, predicGBC)*100))
#print("Accuracy con Datos validación {:.2f}".format(accuracy_score(y_val, predicGBC)*100))
r2 = r2_score(y_test_hsA, predicGBC)
r2

La probabilidad de acierto  para GradientBoostingClassifier 

67.03960617587828
1067
Matrix de Confusion 

Predicted     0     1   All
True                       
0          2280   351  2631
1          1122   716  1838
All        3402  1067  4469
              precision    recall  f1-score   support

           0       0.67      0.87      0.76      2631
           1       0.67      0.39      0.49      1838

    accuracy                           0.67      4469
   macro avg       0.67      0.63      0.62      4469
weighted avg       0.67      0.67      0.65      4469

Accuracy con Datos test 67.04


-0.3612777509637537

In [42]:
from sklearn.naive_bayes import GaussianNB
NB = GaussianNB()
NB.fit(X_train, y_train_hsA)
predicNB = NB.predict(X_test)

print('la probabilidad de acierto  Naiver Bayes\n')
print((np.sum(predicNB == y_test_hsA)/len(predicNB))*100)
print(predicNB.sum())
print('matrix de Confusion')
print(pd.crosstab(y_test_hsA, predicNB, rownames=['True'], colnames=['Predicted'], margins=True))
print(classification_report(y_test_hsA, predicNB))

la probabilidad de acierto  Naiver Bayes

64.55582904452898
2606
matrix de Confusion
Predicted     0     1   All
True                       
0          1455  1176  2631
1           408  1430  1838
All        1863  2606  4469
              precision    recall  f1-score   support

           0       0.78      0.55      0.65      2631
           1       0.55      0.78      0.64      1838

    accuracy                           0.65      4469
   macro avg       0.66      0.67      0.65      4469
weighted avg       0.69      0.65      0.65      4469



In [44]:
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression(C=1e5)
logreg.fit(X_train, y_train_hsA)
prediclog= logreg.predict(X_test)

accuracy_score(y_test_hsA, prediclog)
print('la probabilidad de acierto  Regresión logistica\n')
print((np.sum(prediclog == y_test_hsA)/len(prediclog))*100)
print(prediclog.sum())
print('matrix de Confusion')
print(pd.crosstab(y_test_hsA, prediclog, rownames=['True'], colnames=['Predicted'], margins=True))
print(classification_report(y_test_hsA, prediclog))


la probabilidad de acierto  Regresión logistica

68.49407026180353
1594
matrix de Confusion
Predicted     0     1   All
True                       
0          2049   582  2631
1           826  1012  1838
All        2875  1594  4469
              precision    recall  f1-score   support

           0       0.71      0.78      0.74      2631
           1       0.63      0.55      0.59      1838

    accuracy                           0.68      4469
   macro avg       0.67      0.66      0.67      4469
weighted avg       0.68      0.68      0.68      4469



C:\Users\daalu\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [48]:
SVM = SVC(kernel="linear")
SVM.fit(X_train,y_train_hsA)

predicSVM = SVM.predict(X_test)
#print('combinacion',pair)
print('la probabilidad de acierto Máquina de vectores de soporte Lineal\n')
print((np.sum(predicSVM == y_test_hsA)/len(predicSVM))*100)
print(predicSVM.sum())
print('matrix de Confusion')
print(pd.crosstab(y_test_hsA, predicSVM, rownames=['True'], colnames=['Predicted'], margins=True))
print(classification_report(y_test_hsA, predicSVM))

la probabilidad de acierto Máquina de vectores de soporte Lineal

71.44775117475946
1466
matrix de Confusion
Predicted     0     1   All
True                       
0          2179   452  2631
1           824  1014  1838
All        3003  1466  4469
              precision    recall  f1-score   support

           0       0.73      0.83      0.77      2631
           1       0.69      0.55      0.61      1838

    accuracy                           0.71      4469
   macro avg       0.71      0.69      0.69      4469
weighted avg       0.71      0.71      0.71      4469



In [49]:
SVM = SVC(degree=2, kernel="poly",probability=True)
SVM.fit(X_train,y_train_hsA)

predicSVM = SVM.predict(X_test)
#print('combinacion',pair)
print('la probabilidad de acierto Máquina de vectores de soporte Lineal\n')
print((np.sum(predicSVM == y_test_hsA)/len(predicSVM))*100)
print(predicSVM.sum())
print('matrix de Confusion')
print(pd.crosstab(y_test_hsA, predicSVM, rownames=['True'], colnames=['Predicted'], margins=True))
print(classification_report(y_test_hsA, predicSVM))

la probabilidad de acierto Máquina de vectores de soporte Lineal

64.89147460281941
423
matrix de Confusion
Predicted     0    1   All
True                      
0          2554   77  2631
1          1492  346  1838
All        4046  423  4469
              precision    recall  f1-score   support

           0       0.63      0.97      0.77      2631
           1       0.82      0.19      0.31      1838

    accuracy                           0.65      4469
   macro avg       0.72      0.58      0.54      4469
weighted avg       0.71      0.65      0.58      4469



# Conclusión

El mejor algorimo de clasificación supervisada es el de Naiver Bayes Gaussian, dado que es el unico que no presenta un sobreajuste o uno sin ajuste, no obstante su métrica de presición o accuracy es bajo del $65\%$.  